core/data_array.cpp

```diff
@@ -223,58 +211,88 @@ Variable sparse_dense_op_impl(const VariableConstView &sparseCoord_,
                  }});
 }
 
+namespace {
+bool is_self_append(const DataArrayConstView &a, const DataArrayConstView &b) {
+  return &a.underlying() == &b.underlying();
+}
+} // namespace
+
 DataArray &DataArray::operator+=(const DataArrayConstView &other) {
   expect::coordsAreSuperset(*this, other);
   union_or_in_place(masks(), other.masks());
-  data() += other.data();
+  if (hasData() && other.hasData()) {
+    data() += other.data();
+  } else {
+    if (is_self_append(*this, other))
+      event::append(unaligned(), copy(other.unaligned()));
+    else
+      event::append(unaligned(), other.unaligned());
+  }
   return *this;
 }
 
 DataArray &DataArray::operator-=(const DataArrayConstView &other) {
   expect::coordsAreSuperset(*this, other);
   union_or_in_place(masks(), other.masks());
-  data() -= other.data();
+  if (hasData() && other.hasData()) {
+    data() -= other.data();
+  } else {
+    const DataArray tmp =
+        is_self_append(*this, other) ? copy(other.unaligned()) : DataArray{};
+    auto other_unaligned = tmp ? DataArrayConstView(tmp) : other.unaligned();
+    unaligned().data() *= -1.0;
+    event::append(unaligned(), other_unaligned);
+    unaligned().data() *= -1.0;
+  }
   return *this;
 }
 
+namespace {
+void expect_hist(const DataArrayConstView &b, const Dim dim) {
+  if (!is_histogram(b, dim))
+    throw except::RealignedDataError(
+        "Multiplication/division of realigned data requires histogram has "
+        "second operand.");
+}
+} // namespace
+
 template <class Op>
 DataArray &sparse_dense_op_inplace(Op op, DataArray &a,
-                                   const DataArrayConstView &b, const Dim dim) {
-  if (!is_sparse_and_histogram(a, b, dim)) {
+                                   const DataArrayConstView &b) {
+  if (unaligned::is_realigned_events(a)) {
+    const auto &bins = unaligned::realigned_event_coord(a);
+    const Dim dim = bins.dims().inner();
+    expect_hist(b, dim);
```

Comment:
I don't think any of your new tests cover this case

Reply:
Right, I thought the modified test covers it, but it did not.